### ISSUE TITLE
lockdep warning: possible recursive locking (&tq->tq_lock) at taskq_dispatch in z_rd_int_2

### ISSUE THREAD

**Issue by ofaaland** [zfsonlinux/spl#485]
lockdep reported a different scenario potentially leading to deadlock on tq_lock
```
[ INFO: possible recursive locking detected ]                                      
2.6.32-504.16.2.1chaos.ch5.3.x86_64.debug #1                                       
---------------------------------------------                                      
z_rd_int_2/824 is trying to acquire lock:                                          
 (&tq->tq_lock){..-...}, at: [<ffffffffa01d5dc2>] taskq_dispatch+0x32/0x1a0 [spl]  

but task is already holding lock:
 (&tq->tq_lock){..-...}, at: [<ffffffffa01d6565>] taskq_dispatch_ent+0x35/0x170 [spl]

other info that might help us debug this:
1 lock held by z_rd_int_2/824:           
 #0:  (&tq->tq_lock){..-...}, at: [<ffffffffa01d6565>] taskq_dispatch_ent+0x35/0x170 [spl]

stack backtrace:
Pid: 824, comm: z_rd_int_2 Not tainted 2.6.32-504.16.2.1chaos.ch5.3.x86_64.debug #1
Call Trace:                                                                        
 <IRQ>  [<ffffffff810bfe90>] ? __lock_acquire+0x11b0/0x1560                        
 [<ffffffff81042f68>] ? pvclock_clocksource_read+0x58/0xd0                         
 [<ffffffff810c02e4>] ? lock_acquire+0xa4/0x120                                    
 [<ffffffffa01d5dc2>] ? taskq_dispatch+0x32/0x1a0 [spl]                            
 [<ffffffff810acb48>] ? sched_clock_cpu+0xb8/0x110                                 
 [<ffffffffa01d6be0>] ? taskq_thread_spawn_task+0x0/0x40 [spl]                     
 [<ffffffff81561d95>] ? _spin_lock_irqsave+0x55/0xa0                               
 [<ffffffffa01d5dc2>] ? taskq_dispatch+0x32/0x1a0 [spl]                            
 [<ffffffffa0311d40>] ? zio_execute+0x0/0x180 [zfs]                                
 [<ffffffffa01d5dc2>] ? taskq_dispatch+0x32/0x1a0 [spl]                            
 [<ffffffffa0311d40>] ? zio_execute+0x0/0x180 [zfs]                                
 [<ffffffffa01d5f92>] ? taskq_thread_spawn+0x62/0x70 [spl]                         
 [<ffffffffa01d65ef>] ? taskq_dispatch_ent+0xbf/0x170 [spl]                        
 [<ffffffffa0311d40>] ? zio_execute+0x0/0x180 [zfs]                                
 [<ffffffffa0298b6e>] ? spa_taskq_dispatch_ent+0x7e/0xb0 [zfs]                     
 [<ffffffffa030d583>] ? zio_taskq_dispatch+0x73/0xa0 [zfs]                         
 [<ffffffffa02c2ef1>] ? vdev_disk_dio_free+0x51/0x60 [zfs]                         
 [<ffffffffa030d5e5>] ? zio_interrupt+0x15/0x20 [zfs]                              
 [<ffffffffa02c2f79>] ? vdev_disk_dio_put+0x79/0x90 [zfs]                          
 [<ffffffffa02c345c>] ? vdev_disk_physio_completion+0x2c/0x70 [zfs]                
 [<ffffffff811e6fed>] ? bio_endio+0x1d/0x40                                        
 [<ffffffff8129241b>] ? req_bio_endio+0x9b/0xe0                                    
 [<ffffffff81293b3c>] ? blk_update_request+0x11c/0x520                             
 [<ffffffff81293df9>] ? blk_update_request+0x3d9/0x520                             
 [<ffffffff813b3cff>] ? scsi_device_unbusy+0xbf/0xe0                               
 [<ffffffff81041ffc>] ? kvm_clock_read+0x1c/0x20                                   
 [<ffffffff81293f67>] ? blk_update_bidi_request+0x27/0xa0                          
 [<ffffffff8129606f>] ? blk_end_bidi_request+0x2f/0x80                             
 [<ffffffff81296110>] ? blk_end_request+0x10/0x20                                  
 [<ffffffff813b3f20>] ? scsi_io_completion+0xc0/0x6c0                              
 [<ffffffff813aa14d>] ? scsi_finish_command+0xcd/0x130                             
 [<ffffffff813b4687>] ? scsi_softirq_done+0x147/0x170                              
 [<ffffffff8129c22d>] ? blk_done_softirq+0x8d/0xa0                                 
 [<ffffffff8108215f>] ? __do_softirq+0xdf/0x210
 [<ffffffff8100c40c>] ? call_softirq+0x1c/0x30
 [<ffffffff8100fd1d>] ? do_softirq+0xad/0xe0
 [<ffffffff81081e95>] ? irq_exit+0x95/0xa0
 [<ffffffff815691a5>] ? do_IRQ+0x75/0xf0
 [<ffffffff8100ba53>] ? ret_from_intr+0x0/0x16
 <EOI>  [<ffffffffa0294093>] ? SHA256Transform+0x133/0x220 [zfs]
 [<ffffffffa0294230>] ? zio_checksum_SHA256+0xb0/0x210 [zfs]
 [<ffffffff810bdbdd>] ? lock_release_holdtime+0x3d/0x190
 [<ffffffff810baabd>] ? trace_hardirqs_off+0xd/0x10
 [<ffffffff81561a07>] ? _spin_unlock_irqrestore+0x67/0x80
 [<ffffffffa0314cb8>] ? zio_checksum_error+0x198/0x680 [zfs]
 [<ffffffffa030da52>] ? zio_remove_child+0x1b2/0x260 [zfs]
 [<ffffffffa030dbbe>] ? zio_checksum_verify+0x2e/0xa0 [zfs]
 [<ffffffffa030d2c4>] ? zio_wait_for_children+0x134/0x140 [zfs]
 [<ffffffffa031109f>] ? zio_vdev_io_assess+0x2f/0x270 [zfs]
 [<ffffffffa0313f00>] ? zio_done+0xb50/0x14b0 [zfs]
 [<ffffffffa0311e18>] ? zio_execute+0xd8/0x180 [zfs]
 [<ffffffffa01d6221>] ? taskq_thread+0x281/0x590 [spl]
 [<ffffffff81066f00>] ? default_wake_function+0x0/0x20
 [<ffffffffa01d5fa0>] ? taskq_thread+0x0/0x590 [spl]
 [<ffffffff810a4cbe>] ? kthread+0x9e/0xc0
 [<ffffffff8100c30a>] ? child_rip+0xa/0x20
 [<ffffffff8100bb10>] ? restore_args+0x0/0x30
 [<ffffffff810a4c20>] ? kthread+0x0/0xc0
 [<ffffffff8100c300>] ? child_rip+0x0/0x20

**Comment by ofaaland** [comment]
Is this a duplicate of https://github.com/zfsonlinux/spl/issues/483 ?

**Comment by behlendorf** [comment]
Closing.  Resolved by 326172d8549e0a34a8e4ef4665d8bdfcf7aeda6e.
